core/src/main/scala/cats/data/ContT.scala

```diff
@@ -0,0 +1,108 @@
+package cats
+package data
+
+/**
+ * This is a continuation transformer based on the ContT in
+ * the Haskell package Control.Monad.Cont
+ *
+ * This is reasonably straight-forward except that to make
+ * a tailRecM implementation we leverage the Defer type class to
+ * obtain stack-safety.
+ */
+sealed trait ContT[M[_], A, B] extends Serializable { self =>
+  final def run: (B => M[A]) => M[A] = runAndThen
+  protected def runAndThen: AndThen[B => M[A], M[A]]
+
+  def map[C](fn: B => C): ContT[M, A, C] = {
+    // allocate/pattern match once
+    val fnAndThen = AndThen(fn)
+    ContT { fn2 =>
+      run(fnAndThen.andThen(fn2))
+    }
+  }
+
+  /**
+   * c.mapCont(f).run(g) == f(c.run(g))
+   */
+  def mapCont(fn: M[A] => M[A]): ContT[M, A, B] =
+    // Use later here to avoid forcing run
+    ContT.later(runAndThen.andThen(fn))
+
+  /**
+   * cont.withCont(f).run(cb) == cont.run(f(cb))
+   */
+  def withCont[C](fn: (C => M[A]) => B => M[A]): ContT[M, A, C] =
+    // lazy to avoid forcing run
+    ContT.later(AndThen(fn).andThen(runAndThen))
+
+  def flatMap[C](fn: B => ContT[M, A, C]): ContT[M, A, C] = {
+    // allocate/pattern match once
+    val fnAndThen = AndThen(fn)
+    ContT[M, A, C] { fn2 =>
+      val contRun: ContT[M, A, C] => M[A] = (_.run(fn2))
+      val fn3: B => M[A] = fnAndThen.andThen(contRun)
+      run(fn3)
```

Comment:
I wonder if we should take `(implicit M: Defer[M])` and do `M.defer(run(fn3))` here. If we do this, I think it is possibly that this will be a stack safe Monad without a specialized tailRecM. I'll experiment with that.